csharp/ql/test/query-tests/API Abuse/DisposeNotCalledOnException/DisposeNotCalledOnException.cs

```diff
@@ -58,6 +58,10 @@ public void Method()
         SqlConnection c1f = new SqlConnection();
         Throw2(c1f);
         c1f.Dispose();
+
+        // GOOD: using declaration
+        using SqlConnection c2 = new SqlConnection("");
+        c2.Open();
```

Comment:
I wonder if the query should be able to handle:
```
        // GOOD: using declaration
        using SqlConnection c3 = new SqlConnection("");
        Throw2(c3);
        c3.Dispose();
```
It currently incorrectly flags this.

One way to fix this is to introduce a
```
private predicate localFlowStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
  DataFlow::localFlowStep(nodeFrom, nodeTo) and
  not exists(AssignableDefinition def, UsingDeclStmt uds |
    nodeTo.asExpr() = def.getAReachableRead() and
    def.getTargetAccess() = uds.getAVariableDeclExpr().getAccess()
  )
}
```
in `DisposeNotCalledOnException.ql` and replace `DataFlow::localFlowStep` with `localFlowStep` in the predicate `reachesDisposeCall`.

Reply:
I'm not convinced that we should be handling this case. It would be a double disposal error. It would be like
```
using(var c3 = new SqlConnection())
{
  Throw2(c3);
  c3.Dispose();
}
```
The `Dispose` call makes no sense.

Reply:
Yeah, I know, but according to [the spec](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable.dispose?view=netframework-4.8), it is OK to call `Dispose` multiple times:
```
If an object's Dispose method is called more than once, the object must ignore all calls after the first one. 
```
so I wonder if code like this could happen in practice:
```
void M()
{
    using var x = new FileStream(...);
    if (...) {
       // do stuff with x
       x.Dispose(); // early dispose to
       // do lengthy computations
    }
    else
    {
        // do stuff with x
     }
}
```

Reply:
Very interesting. This is also flagged by the current query:
```
        using(SqlConnection c3 = new SqlConnection(""))
        {
            Throw2(c3);
            c3.Dispose();
        }
```
Perhaps the change you suggest could be made against `master` on a different PR, as it's not specific to the newly introduced `using` declarations.

Reply:
OK, let's do it on a separate PR.